p2p/protocol/circuitv2/relay/relay.go

```diff
@@ -105,17 +105,24 @@ func New(h host.Host, opts ...Option) (*Relay, error) {
 
 func (r *Relay) Close() error {
 	r.mx.Lock()
-	defer r.mx.Unlock()
 	if !r.closed {
 		r.closed = true
+		r.mx.Unlock()
+
 		r.host.RemoveStreamHandler(proto.ProtoIDv2Hop)
 		r.host.Network().StopNotify(r.notifiee)
 		r.scope.Done()
 		r.cancel()
+
+		r.mx.Lock()
```

Comment:
Would it make sense to move the untagging a few lines up? That way you'd only need to lock the mutex once.

Reply:
That line actually has a race condition :( 
if both r.gc and r.Close(waiting on lock before untag peer) happen at the same time we'd untag the peer twice with the conn manager since in close we don't delete the entry from the map. 

I updated it to also cleanup the background goroutine like it happens in other services. If you want I'll make the exact change you asked.